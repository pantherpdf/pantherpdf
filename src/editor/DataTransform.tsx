/**
 * DataTransform.tsx
 * Manage data transformations. Add, delete, edit Transform.
 */

import React, { useState } from 'react';
import { Report, TransformItem } from '../types';
import { GeneralProps, Transform } from './types';
import { Dropdown, Button, ButtonGroup, Modal } from 'react-bootstrap';
import Trans, { TransName } from '../translation';
import { getTransform } from '../transforms/allTransforms';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faArrowDown,
  faArrowUp,
  faEdit,
  faFolderOpen,
  faPlus,
  faTimes,
  faTrash,
} from '@fortawesome/free-solid-svg-icons';
import ObjectExplorer from './ObjectExplorer';
import InputApplyOnEnter from '../widgets/InputApplyOnEnter';
import globalStyle from '../globalStyle.module.css';

/**
 * Function for calling all transformations
 * @func
 * @param {inputData} unknown - Input data
 * @param {len} number - Number of transformations to apply
 */
export async function transformData(
  allTrans: Transform[],
  inputData: unknown,
  transformData: TransformItem[],
  len?: number,
) {
  if (len === undefined) {
    len = transformData.length;
  }
  for (let i = 0; i < len; ++i) {
    const w = transformData[i];
    const comp = getTransform(allTrans, w.type);
    inputData = await comp.transform(inputData, w);
  }
  return inputData;
}

interface TransformItemProps extends GeneralProps {
  item: TransformItem;
  index: number;
  showData: (len: number) => void;
  up: (idx: number) => void;
  down: (idx: number) => void;
  itemDelete: (idx: number) => void;
  openEditor: (idx: number) => void;
}
function TransformItemEditor(props: TransformItemProps) {
  const comp = getTransform(props.transforms, props.item.type);
  return (
    <Dropdown as={ButtonGroup} className="d-flex" size="sm">
      <button
        className="btn btn-sm btn-outline-secondary flex-fill text-left"
        onClick={() => props.showData(props.index + 1)}
      >
        <span className="d-block">{TransName(comp.name)}</span>
        <small
          className="d-block text-truncate text-nowrap text-muted"
          style={{ maxWidth: '10rem' }}
        >
          {props.item.comment}
        </small>
      </button>
      <Dropdown.Toggle
        style={{ maxWidth: '3rem' }}
        split
        variant="outline-secondary"
        id="dropdown-custom-2"
      />
      <Dropdown.Menu>
        <Dropdown.Item onClick={() => props.openEditor(props.index)}>
          <FontAwesomeIcon icon={faEdit} className="me-2" fixedWidth />
          {Trans('edit')}
        </Dropdown.Item>
        <Dropdown.Item
          onClick={e => {
            e.preventDefault();
            props.up(props.index);
          }}
          disabled={props.index === 0}
        >
          <FontAwesomeIcon icon={faArrowUp} className="me-2" fixedWidth />
          {Trans('up')}
        </Dropdown.Item>
        <Dropdown.Item
          onClick={e => {
            e.preventDefault();
            props.down(props.index);
          }}
          disabled={props.index + 1 === props.report.transforms.length}
        >
          <FontAwesomeIcon icon={faArrowDown} className="me-2" fixedWidth />
          {Trans('down')}
        </Dropdown.Item>
        <Dropdown.Divider />
        <Dropdown.Item
          onClick={e => {
            e.preventDefault();
            props.itemDelete(props.index);
          }}
        >
          <FontAwesomeIcon icon={faTrash} className="me-2" fixedWidth />
          {Trans('delete')}
        </Dropdown.Item>
      </Dropdown.Menu>
    </Dropdown>
  );
}

async function uploadOverrideSourceData(): Promise<unknown> {
  return new Promise(resolve => {
    let el = document.createElement('INPUT') as HTMLInputElement;
    el.type = 'file';
    el.accept = 'application/json';
    el.multiple = false;
    el.addEventListener('change', function (ev2) {
      const f = el.files && el.files.length > 0 ? el.files[0] : undefined;
      if (!f) {
        return resolve(undefined);
      }
      try {
        const reader = new FileReader();
        reader.onload = e2 => {
          if (!e2.target || typeof e2.target.result !== 'string') {
            return;
          }
          const dt = JSON.parse(e2.target.result);
          resolve(dt);
        };
        reader.readAsText(f);
      } catch (e) {
        alert(`Error: ${String(e)}`);
        return resolve(undefined);
      }
    });
    el.click();
  });
}

interface TEdit {
  index: number;
  data: TransformItem;
}

interface SourceDataShow {
  length?: number;
  data: unknown;
  errorMsg?: string;
}

export default function DataTransform(props: GeneralProps) {
  const [shownModalInsert, setShownModalInsert] = useState<boolean>(false);
  const [showEdit, setShowEdit] = useState<TEdit | null>(null);
  const [showData, setShowData] = useState<SourceDataShow | null>(null);

  async function doShowData(len: number) {
    let dt;
    try {
      dt = await props.getSourceData();
    } catch (e) {
      let msg = String(e);
      if (msg.trim().length === 0) {
        msg = 'unknown error';
      }
      setShowData({ data: null, errorMsg: msg });
      return;
    }

    let dt2;
    try {
      dt2 = await transformData(
        props.transforms,
        dt,
        props.report.transforms,
        len,
      );
    } catch (e) {
      let msg = String(e);
      if (msg.trim().length === 0) {
        msg = 'unknown error';
      }
      setShowData({ data: null, errorMsg: msg });
      return;
    }
    setShowData({ data: dt2, length: len });
  }

  async function itemAdd(key: string) {
    const cmp = getTransform(props.transforms, key);
    const dt = await cmp.newItem();
    const report2: Report = {
      ...props.report,
      transforms: [...props.report.transforms, dt],
    };
    await props.setReport(report2);
    setShownModalInsert(false);
    setShowEdit({
      index: report2.transforms.length - 1,
      data: JSON.parse(JSON.stringify(dt)),
    });
  }

  async function itemUp(idx: number) {
    const report2: Report = {
      ...props.report,
      transforms: [...props.report.transforms],
    };
    const tmp = report2.transforms[idx - 1];
    report2.transforms[idx - 1] = report2.transforms[idx];
    report2.transforms[idx] = tmp;
    return props.setReport(report2);
  }

  async function itemDown(idx: number) {
    const report2: Report = {
      ...props.report,
      transforms: [...props.report.transforms],
    };
    const tmp = report2.transforms[idx + 1];
    report2.transforms[idx + 1] = report2.transforms[idx];
    report2.transforms[idx] = tmp;
    return props.setReport(report2);
  }

  async function itemDelete(idx: number) {
    const report2: Report = {
      ...props.report,
      transforms: [...props.report.transforms],
    };
    report2.transforms.splice(idx, 1);
    return props.setReport(report2);
  }

  function itemOpenEditor(idx: number) {
    setShowEdit({
      index: idx,
      data: JSON.parse(JSON.stringify(props.report.transforms[idx])),
    });
  }

  async function editSave() {
    if (!showEdit) {
      return;
    }
    const report2: Report = {
      ...props.report,
      transforms: [...props.report.transforms],
    };
    report2.transforms[showEdit.index] = showEdit.data;
    await props.setReport(report2);
    setShowEdit(null);
  }

  function editChange(data: TransformItem) {
    if (showEdit) {
      setShowEdit({ ...showEdit, data });
    }
  }

  const editCmp = showEdit
    ? getTransform(props.transforms, showEdit.data.type)
    : null;

  return (
    <>
      <div className={`${globalStyle.section} d-flex`}>
        <label htmlFor="source-url" className="flex-fill">
          {Trans('source data')}
        </label>
        {!!props.setSourceDataOverride && (
          <button
            className="btn btn-link py-0"
            style={{ color: '#666' }}
            onClick={async () => {
              if (!props.setSourceDataOverride) {
                return;
              }
              if (props.isSourceDataOverriden) {
                // remove
                props.setSourceDataOverride(undefined);
              } else {
                // select file and replace
                const dt = await uploadOverrideSourceData();
                if (dt) {
                  props.setSourceDataOverride(dt);
                }
              }
            }}
            title={Trans('load local json file')}
          >
            <FontAwesomeIcon
              icon={props.isSourceDataOverriden ? faTimes : faFolderOpen}
            />
          </button>
        )}
      </div>
      <InputApplyOnEnter
        id="source-url"
        value={props.report.dataUrl}
        onChange={val => {
          const report2: Report = { ...props.report, dataUrl: String(val) };
          props.setReport(report2);
        }}
        placeholder="https://www.example.com/data.js(on)"
      />

      <div className={globalStyle.section}>{Trans('transform')}</div>
      {props.report.transforms.map((item, idx) => (
        <TransformItemEditor
          {...props}
          item={item}
          index={idx}
          key={idx}
          showData={doShowData}
          up={itemUp}
          down={itemDown}
          itemDelete={itemDelete}
          openEditor={itemOpenEditor}
        />
      ))}
      <div className="">
        <button
          className="btn btn-sm btn-outline-secondary mt-2"
          onClick={() => setShownModalInsert(true)}
        >
          <FontAwesomeIcon icon={faPlus} className="me-2" />
          {Trans('transform')}
        </button>
      </div>

      {/* ADD NEW */}
      <Modal show={shownModalInsert} onHide={() => setShownModalInsert(false)}>
        <Modal.Header closeButton>
          <Modal.Title>{Trans('insert data transform')}</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <div className="list-group">
            {props.transforms.map(w => {
              return (
                <button
                  className="btn list-group-item list-group-item-action"
                  key={w.id}
                  onClick={() => itemAdd(w.id)}
                >
                  {TransName(w.name)}
                </button>
              );
            })}
          </div>
        </Modal.Body>
      </Modal>

      {/* EDIT */}
      <Modal show={!!showEdit} onHide={() => setShowEdit(null)}>
        {editCmp && showEdit && (
          <>
            <Modal.Header closeButton>
              <Modal.Title>
                {Trans('edit')} {TransName(editCmp.name)}
              </Modal.Title>
            </Modal.Header>
            <Modal.Body>
              <editCmp.RenderEditor
                {...props}
                index={showEdit.index}
                item={showEdit.data}
                setItem={editChange}
              />
            </Modal.Body>
            <Modal.Footer>
              <input
                type="text"
                className="form-control flex-fill w-auto"
                id="trans-edit-comment"
                value={showEdit.data.comment}
                placeholder={Trans('comment')}
                onChange={e =>
                  editChange({
                    ...showEdit.data,
                    comment: e.currentTarget.value,
                  })
                }
              />
              <Button variant="secondary" onClick={() => setShowEdit(null)}>
                {Trans('cancel')}
              </Button>
              <Button variant="secondary" onClick={editSave}>
                {Trans('save')}
              </Button>
            </Modal.Footer>
          </>
        )}
      </Modal>

      {/* DATA */}
      <Modal show={!!showData} onHide={() => setShowData(null)}>
        <Modal.Header closeButton>
          <Modal.Title>{Trans('data')}</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <>
            {showData?.data && <ObjectExplorer data={showData.data} />}
            {showData?.errorMsg && <div>{showData?.errorMsg}</div>}
          </>
        </Modal.Body>
      </Modal>
    </>
  );
}
