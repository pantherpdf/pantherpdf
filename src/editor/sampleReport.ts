/**
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2023
 * @license MIT
 */

import type { Report } from '../types';

export const sampleReport: Report = {
  name: 'Report 1',
  widgets: [],
  transforms: [],
  properties: {},
  variables: [],
};
