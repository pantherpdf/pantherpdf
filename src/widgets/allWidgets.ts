/**
 * allWidgets.tsx
 * list of all widgets
 */


import type { Widget } from '../editor/types'
import { TextSimple } from './TextSimple'
import { Repeat } from './Repeat'
import { TextHtml } from './TextHtml'
import { Image } from './Image'
import { Columns, ColumnsCt } from './Columns'
import { Condition } from './Condition'
import { FirstMatch } from './FirstMatch'
import { Html } from './Html'
import { Separator } from './Separator'
import { Spacer } from './Spacer'
import { PageBreak } from './PageBreak'
import { Frame } from './Frame'
import { SetVar } from './SetVar'
import { UpdateVar } from './UpdateVar'

export const allWidgets: {[key: string]: Widget} = {
	TextSimple,
	TextHtml,
	Repeat,
	Image,
	Columns,
	ColumnsCt,
	Condition,
	FirstMatch,
	Html,
	Separator,
	Spacer,
	PageBreak,
	Frame,
	SetVar,
	UpdateVar,
}

export default function getWidget(type: string): Widget {
	const c = allWidgets[type]
	if (c !== undefined)
		return c
	throw new Error(`Missing widget ${type}`)
}
