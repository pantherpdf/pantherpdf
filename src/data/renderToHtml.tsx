/**
 * @file Compiled report -> html
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2021-2024
 * @license MIT
 */

import React, { CSSProperties } from 'react';
import { renderToString } from 'react-dom/server';
import { defaultReportCss } from '../types';
import type { ApiEndpoints, ReportCompiled } from '../types';
import type { WidgetPreviewPropsBase, Widget } from '../widgets/types';
import { getWidget } from '../widgets/allWidgets';
import { PropertyFontGenCss } from '../widgets/PropertyFont';

export function renderBody(
  report: ReportCompiled,
  widgets: Widget[],
  externalHelpers: { [key: string]: unknown } = {},
): React.ReactElement {
  const helper: WidgetPreviewPropsBase = {
    renderItem: (item, helper) => {
      const w = getWidget(widgets, item.type);
      return w.Preview({ ...helper, item });
    },
    renderChildren: (chs, helper) => {
      if (!Array.isArray(chs)) {
        throw new Error('Bad childs array');
      }
      return chs.map((item, idx) => {
        const w = getWidget(widgets, item.type);
        return (
          <React.Fragment key={idx}>
            {w.Preview({ ...helper, item })}
          </React.Fragment>
        );
      });
    },
    externalHelpers,
  };

  const children = helper.renderChildren(report.widgets, helper);

  // prepare css
  const cssObj: CSSProperties = {
    ...defaultReportCss,
    ...PropertyFontGenCss(report.properties.font || {}),
  };
  return <body style={cssObj}>{children}</body>;
}

export default function renderToHtml(
  report: ReportCompiled,
  widgets: Widget[],
  api: ApiEndpoints,
  externalHelpers: { [key: string]: unknown } = {},
): string {
  // render content
  const bodyElement = renderBody(report, widgets, externalHelpers);
  const bodyTxt = renderToString(bodyElement);

  const fontUrl = api.fonts?.getCssUrls(report.properties.fontsUsed) || [];
  const fontHtml = fontUrl
    .map(url => `<link rel="stylesheet" href="${url}"></link>`)
    .join('\n');

  const html = `<!DOCTYPE html>
<html lang="en-US">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<title>Report</title>
<style>
/* http://meyerweb.com/eric/tools/css/reset/ 
   v2.0 | 20110126
   License: none (public domain)
*/
html, body, div, span, applet, object, iframe,
h1, h2, h3, h4, h5, h6, p, blockquote, pre,
a, abbr, acronym, address, big, cite, code,
del, dfn, em, img, ins, kbd, q, s, samp,
small, strike, strong, sub, sup, tt, var,
b, u, i, center,
dl, dt, dd, ol, ul, li,
fieldset, form, label, legend,
table, caption, tbody, tfoot, thead, tr, th, td,
article, aside, canvas, details, embed, 
figure, figcaption, footer, header, hgroup, 
menu, nav, output, ruby, section, summary,
time, mark, audio, video {
	margin: 0;
	padding: 0;
	border: 0;
	font-size: 100%;
	font: inherit;
	vertical-align: baseline;
}
/* HTML5 display-role reset for older browsers */
article, aside, details, figcaption, figure, 
footer, header, hgroup, menu, nav, section {
	display: block;
}
body {
	line-height: 1.25;
}
ol, ul {
	list-style: none;
}
blockquote, q {
	quotes: none;
}
blockquote:before, blockquote:after,
q:before, q:after {
	content: '';
	content: none;
}
table {
	border-collapse: collapse;
	border-spacing: 0;
}
/* modifications */
b, strong {
	font-weight: bold;
}
cite, em, i {
	font-style: italic;
}
big {
	font-size: larger;
}
small {
	font-size: smaller;
}
mark {
	background-color: yellow;
}
sub {
    vertical-align: sub;
    font-size: smaller;
}
sup {
    vertical-align: super;
    font-size: smaller;
}
</style>
<style>
* {
	-webkit-print-color-adjust: exact;
	print-color-adjust: exact;
}
</style>
<style>
${report.properties.globalCss}
</style>
${fontHtml}
</head>
${bodyTxt}
</html>
`;
  return html;
}
