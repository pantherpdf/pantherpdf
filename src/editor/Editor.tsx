/**
 * Editor.tsx
 * Managing global state, prepare GeneralProps, drag-drop functions
 */

import React, { ReactNode, useEffect, useRef } from 'react';
import { useState } from 'react';
import getWidget from '../widgets/allWidgets';
import { TReport, TData, FileUploadData } from '../types';
import {
  EditorProps,
  GeneralProps,
  NewItemProps,
  TDragObj,
  TSourceData,
} from './types';
import style from './Editor.module.css';
import Layout from './EditorLayout';
import {
  findInList,
  removeFromList,
  insertIntoList,
  updateDestAfterRemove,
  idCmp,
  updateItem,
} from './childrenMgmt';
import { extractFiles } from '../FileSelect';
import { Image as ImageWidget, ImageData } from '../widgets/Image';
import { transformData } from './DataTransform';
import retrieveOriginalSourceData from './retrieveOriginalSourceData';

export function dropImpl(
  report: TReport,
  current: TDragObj,
  dest: number[],
  copy: boolean,
): TReport | null {
  if (dest.length === 0) {
    throw new Error('dest does not exist');
  }

  let toInsert: TData | TData[];
  let report2: TReport = report;
  if (current.type === 'wid') {
    const dragObj3 = current.wid;
    if (copy) {
      toInsert = findInList(report2, dragObj3);
      toInsert = JSON.parse(JSON.stringify(toInsert));
    }
    // calc destination id
    else {
      // detect if dragging item to one of its childs
      let dest2;
      try {
        dest2 = updateDestAfterRemove(dest, dragObj3);
      } catch (e) {
        return null;
      }
      toInsert = findInList(report2, dragObj3);
      // remove from list
      report2 = removeFromList(report2, dragObj3);
      // update destination id
      dest = dest2;
    }
  }
  //
  else if (current.type === 'widget') {
    toInsert = current.widget;
  }
  //
  else if (current.type === 'widgets') {
    toInsert = current.widgets;
  }
  //
  else {
    throw new Error('unknown dragObj2 type');
  }

  // insert
  if (!Array.isArray(toInsert)) {
    report2 = insertIntoList(report2, dest, toInsert);
  } else {
    for (const toInsert2 of toInsert) {
      report2 = insertIntoList(report2, dest, toInsert2);
      dest[dest.length - 1] += 1;
    }
  }

  return report2;
}

export default function Editor(props: EditorProps) {
  const [selected, setSelected] = useState<number[] | null>(null);
  const [data, setData] = useState<TSourceData>({ data: undefined });
  const [overrideSourceData, setSourceDataOverride] = useState<
    string | undefined
  >(undefined);
  const dragObj = useRef<TDragObj | null>(null);

  // refresh data
  useEffect(() => {
    (async function () {
      try {
        const dt1 = await getOrigSourceInternal();
        const dt2 = await transformData(dt1, props.report.transforms, true);
        setData({ data: dt2 });
      } catch (e) {
        setData({ data: undefined, errorMsg: String(e) });
      }
    })();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [overrideSourceData, props.report.dataUrl, props.report.transforms]);

  async function getOrigSourceInternal(): Promise<unknown> {
    if (typeof props.sourceData !== 'undefined') {
      return props.sourceData;
    }
    if (overrideSourceData !== undefined) {
      return JSON.parse(overrideSourceData);
    }
    return retrieveOriginalSourceData({
      reportDataUrl: props.report.dataUrl,
      api: props.api,
      allowUnsafeJsEval: true,
    });
  }

  let props2: GeneralProps;

  async function drop(
    e: React.DragEvent<HTMLDivElement>,
    dest: number[],
  ): Promise<void> {
    e.preventDefault();
    e.stopPropagation();
    e.currentTarget.classList.remove(style.dragging);

    // widget
    if (dragObj.current) {
      const copy = e.altKey || e.metaKey;
      // dragObj.current is empty when you drag-drop some other element like logo, menu ...
      const report2 = dropImpl(props.report, dragObj.current, dest, copy);
      if (report2) {
        props.setReport(report2);
        setSelected(null);
      }
      dragObj.current = null;
      return;
    }

    // file upload
    const files = extractFiles(e.dataTransfer);
    if (files.length > 0) {
      if (!props.api.filesUpload) {
        alert('File upload not supported');
        return;
      }
      const f = files[0];
      const dt: FileUploadData = {
        name: f.name,
        modifiedTime:
          new Date(f.lastModified).toISOString().substring(0, 19) + 'Z',
        mimeType: f.type,
      };
      try {
        await props.api.filesUpload(f, dt, () => {});
      } catch (e) {
        alert(`Error while uploading: ${String(e)}`);
      }
      const newItemProps: NewItemProps = { report: props.report };
      const img = (await ImageWidget.newItem(newItemProps)) as ImageData;
      img.url = `local/${f.name}`;
      const report2 = dropImpl(
        props.report,
        { type: 'widget', widget: img },
        dest,
        false,
      );
      if (report2) {
        props.setReport(report2);
        setSelected(null);
      }
      dragObj.current = null;
      return;
    }
  }

  function dragOver(e: React.DragEvent<HTMLDivElement>): void {
    // widget
    if (dragObj.current) {
      e.stopPropagation();
      e.preventDefault();
      if (dragObj.current.type === 'wid') {
        const copy = e.altKey || e.metaKey;
        e.dataTransfer.dropEffect = copy ? 'copy' : 'move';
      } else {
        e.dataTransfer.dropEffect = 'copy';
      }
      return;
    }

    // file upload
    if (e.dataTransfer.items.length > 0 || e.dataTransfer.files.length > 0) {
      e.stopPropagation();
      e.preventDefault();
      e.dataTransfer.dropEffect = 'copy';
      return;
    }
  }

  function dragEnter(e: React.DragEvent<HTMLDivElement>): void {
    if (dragObj.current) {
      e.currentTarget.classList.add(style.dragging);
    }
  }

  function dragLeave(e: React.DragEvent<HTMLDivElement>): void {
    e.currentTarget.classList.remove(style.dragging);
  }

  function renderSpacer(parents: number[]): ReactNode {
    return (
      <div
        className={style.spacer}
        onDrop={e => drop(e, parents)}
        onDragOver={dragOver}
        onDragEnter={dragEnter}
        onDragLeave={dragLeave}
      ></div>
    );
  }

  function renderWidget(child: TData, wid: number[]): ReactNode {
    const obj = getWidget(child.type);

    return (
      <div
        className={`${style.widgetBox} ${
          (selected && idCmp(selected, wid) && style.selected) || ''
        }`}
        onClick={e => widgetMouseClick(e, wid)}
        draggable={typeof obj.canDrag === 'undefined' || obj.canDrag}
        onDragStart={e => dragWidgetStart(e, { type: 'wid', wid })}
        onDragEnd={e => dragWidgetEnd(e)}
      >
        <obj.Render
          {...props2}
          item={child}
          setItem={(itm: TData) => {
            const r2 = updateItem(props2.report, wid, itm);
            return props2.setReport(r2);
          }}
          wid={wid}
        />
      </div>
    );
  }

  function renderWidgets(children: TData[], parents: number[]): ReactNode {
    return (
      <>
        {renderSpacer([...parents, 0])}
        {children.map((c, idx) => (
          <React.Fragment key={idx}>
            {renderWidget(c, [...parents, idx])}
            {renderSpacer([...parents, idx + 1])}
          </React.Fragment>
        ))}
      </>
    );
  }

  function dragWidgetStart(
    e: React.DragEvent<HTMLDivElement>,
    dragObj2: TDragObj,
  ): void {
    e.stopPropagation();
    dragObj.current = dragObj2;
  }

  function dragWidgetEnd(e: React.DragEvent<HTMLDivElement>): void {
    e.stopPropagation();
    dragObj.current = null;
  }

  function widgetMouseClick(
    e: React.MouseEvent<HTMLDivElement, MouseEvent>,
    wid: number[],
  ) {
    e.preventDefault();
    e.stopPropagation();
    wid = [...wid];
    while (wid.length > 0) {
      const item = findInList(props.report, wid);
      const obj = getWidget(item.type);
      if (typeof obj.canSelect === 'undefined' || obj.canSelect) {
        setSelected(wid);
        break;
      }
      wid.splice(wid.length - 1, 1);
    }
  }

  props2 = {
    ...props,

    selected,
    setSelected,
    ...(typeof props.sourceData === 'undefined'
      ? {
          setSourceDataOverride: (data: unknown) => {
            const dt =
              typeof data !== 'undefined' ? JSON.stringify(data) : undefined;
            setSourceDataOverride(dt);
          },
        }
      : {}),
    getSourceData: getOrigSourceInternal,
    isSourceDataOverriden: !!overrideSourceData,
    data,

    renderWidget,
    renderWidgets,
    dragWidgetStart,
    dragWidgetEnd,
    drop,
  };

  return <Layout {...props2} dragOver={dragOver} />;
}
