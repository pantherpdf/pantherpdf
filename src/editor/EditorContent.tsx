/**
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2023
 * @license MIT
 */

import React, { useEffect } from 'react';
import Typography from '@mui/material/Typography';
import { defaultReportCss } from '../types';
import { propertyFontGenCss } from '../widgets/PropertyFont';
import type { GeneralProps } from './types';
import trans from '../translation';
import getFontsUsed from '../data/getFontsUsed';

export default function EditorContent(props: GeneralProps) {
  useEffect(() => {
    const fontsUsed = getFontsUsed(props.report, props.widgets);
    const urls = props.api.fonts?.getCssUrls(fontsUsed) || [];
    loadFonts(urls);
  });
  const style = {
    ...defaultReportCss,
    ...propertyFontGenCss(props.report.properties.font || {}),
  };
  const width = props.report.properties.paper?.width || 210;
  style.maxWidth = `${(width * 800) / 210}px`;
  style.margin = `0 auto`;
  return (
    <div style={style}>
      {props.renderWidgets(props.report.widgets, [])}
      {props.report.widgets.length === 0 && (
        <Typography color="GrayText" align="center">
          {trans('drag drop widgets here')}
        </Typography>
      )}
    </div>
  );
}

const loadedUrls: string[] = [];
function loadFonts(urls: string[]): void {
  for (const url of urls) {
    if (loadedUrls.indexOf(url) !== -1) {
      continue;
    }
    loadedUrls.push(url);
    const link = window.document.createElement('link');
    link.rel = 'stylesheet';
    link.href = url;
    window.document.head.appendChild(link);
  }
}
