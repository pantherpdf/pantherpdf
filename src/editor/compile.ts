/**
 * @file Compile sourceData and report - evaluate all formulas
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2021
 * @license MIT
 */

import { Report, Item, ApiEndpoints } from '../types';
import type { ReportCompiled } from '../types';
import FormulaEvaluate from '../formula/formula';
import { getWidget } from '../widgets/allWidgets';
import { CompileHelper, Widget } from './types';
import { PropertyFontExtractStyle } from '../widgets/PropertyFont';

type TOvrr = [string, unknown];

export class FormulaHelper {
  overrides: TOvrr[];

  constructor() {
    this.overrides = [];
  }

  push(name: string, valueOrCb: unknown) {
    this.overrides.push([name, valueOrCb]);
  }

  pop() {
    this.overrides.pop();
  }

  async getVar(name: string): Promise<unknown> {
    for (let i = this.overrides.length - 1; i >= 0; --i) {
      if (this.overrides[i][0] === name) {
        let val = this.overrides[i][1];
        if (typeof val == 'function') {
          val = await val(name);
        }
        return val;
      }
    }
    return undefined;
  }
}

export default async function compile(
  report: Report,
  data: unknown,
  widgets: Widget[],
  api: ApiEndpoints,
  externalHelpers: { [key: string]: any } = {},
): Promise<ReportCompiled> {
  // make a copy, to support changing
  report = JSON.parse(JSON.stringify(report));

  const formulaHelper = new FormulaHelper();
  const getVar = formulaHelper.getVar.bind(formulaHelper);
  formulaHelper.push('data', data);
  formulaHelper.push('report', report);

  // custom variables
  const vars: { [key: string]: unknown } = {};
  function getVarValue(varName: string): unknown {
    return vars[varName];
  }
  for (const v of report.variables) {
    vars[v.name] = await FormulaEvaluate(v.formula, { getVar });
    formulaHelper.push(v.name, getVarValue);
  }

  let dt2: ReportCompiled = {
    ...report,
    children: [],
    fontsUsed: [],
    globalCss: '',
  };
  dt2 = JSON.parse(JSON.stringify(dt2));
  if (dt2.properties.font) {
    const style = PropertyFontExtractStyle(dt2.properties.font);
    if (style) {
      dt2.fontsUsed.push(style);
    }
  }

  if (dt2.properties.fileName) {
    const res = await FormulaEvaluate(dt2.properties.fileName, { getVar });
    if (typeof res !== 'string') {
      throw new Error(
        `Filename should be a string, but received ${typeof res}`,
      );
    }
    dt2.properties.fileName = res;
  }

  const helper: CompileHelper = {
    wid: [],
    report: report,
    reportCompiled: dt2,
    formulaHelper,
    api,
    variables: vars,
    externalHelpers,

    evalFormula: async (txt: string) => {
      return FormulaEvaluate(txt, { getVar });
    },

    compileChildren: async (children: Item[], helper: CompileHelper) => {
      // dont use promise.all() because order of execution matters and some async operation could change it
      const arr2 = [];
      for (let i = 0; i < children.length; ++i) {
        const ch = children[i];
        const helper2 = { ...helper, wid: [...helper.wid, i] };
        const dt = await getWidget(widgets, ch.type).compile(ch, helper2);
        arr2.push(dt);
      }
      return arr2;
    },
  };

  dt2.children = await helper.compileChildren(report.children, helper);

  formulaHelper.pop();
  formulaHelper.pop();
  report.variables.map(() => formulaHelper.pop());
  if (formulaHelper.overrides.length !== 0) {
    throw new Error('helper has overrides still left inside');
  }
  return dt2;
}
