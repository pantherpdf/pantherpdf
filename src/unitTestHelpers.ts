/**
 * @file Helper functions used in unit tests
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2023
 * @license MIT
 */

import { Item, ItemCompiled, Report, ReportCompiled } from './types';
import { defaultWidgets } from './widgets/allWidgets';
import compile from './editor/compile';
import { renderToHtmlContent } from './editor/renderToHtml';

export async function compileComponentTest(
  cmpData: object,
  data: unknown,
): Promise<ItemCompiled> {
  const dt: Report = {
    target: 'pdf',
    name: 'John Johnny',
    children: [cmpData as Item],
    transforms: [],
    properties: {},
    dataUrl: '',
    variables: [],
  };
  const reportCompiled = await compile(dt, data, defaultWidgets, {});
  return reportCompiled.children[0];
}

export async function compileTest(
  report: Report,
  data: unknown,
): Promise<ReportCompiled> {
  return compile(report, data, defaultWidgets, {});
}

export function renderToHtmlContentTest(report: ReportCompiled) {
  return renderToHtmlContent(report, defaultWidgets);
}
