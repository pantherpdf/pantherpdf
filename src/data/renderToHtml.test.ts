/**
 * @jest-environment node
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2021-2024
 * @license MIT
 */

import { compileTest, renderWidget } from '../unitTestHelpers';
import type { Report } from '../types';
import type { TextSimpleData } from '../widgets/TextSimple';
import { sampleReport } from '../editor/sampleReport';
import renderToHtml from './renderToHtml';
import { defaultWidgets } from '../widgets/allWidgets';

test('text', async () => {
  const el: TextSimpleData = {
    type: 'TextSimple',
    value: { formula: '"Hello World: "+data.num' },
    children: [],
  };
  const obj = { num: 123 };
  const html = await renderWidget(el, obj);
  expect(html).toMatchSnapshot();
});

test('should include google font', async () => {
  const report: Report = {
    ...sampleReport,
    widgets: [],
  };
  report.properties = {
    ...report.properties,
    font: {
      family: 'Roboto Mono',
    },
  };
  const compiled = await compileTest(report, {});
  const html = renderToHtml(compiled, defaultWidgets);
  const expectedUrl1 =
    'https://fonts.googleapis.com/css2?family=Roboto+Mono:ital,wght@0,400&display=swap';
  const expectedUrl2 =
    'https://fonts.googleapis.com/css2?family=Roboto%20Mono:ital,wght@0,400&display=swap';
  expect(
    html.indexOf(expectedUrl1) !== -1 || html.indexOf(expectedUrl2) !== -1,
  ).toBeTruthy();
});

test('should include globalCss', async () => {
  const compiled = await compileTest(sampleReport, {});
  compiled.properties.globalCss = '.abc-def-123-456 { font-weight: bold }';
  const html = renderToHtml(compiled, defaultWidgets);
  expect(html.indexOf('.abc-def-123-456 { font-weight: bold }')).not.toBe(-1);
});
