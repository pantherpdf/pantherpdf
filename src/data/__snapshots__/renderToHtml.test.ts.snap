// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`text 1`] = `
<div>
  Hello World: 123
</div>
`;
